src/ray/raylet/pubsub/pubsub_client.h

```diff
@@ -0,0 +1,162 @@
+// Copyright 2017 The Ray Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include "ray/common/id.h"
+#include "ray/rpc/worker/core_worker_client_pool.h"
+#include "src/ray/protobuf/common.pb.h"
+
+namespace ray {
+
+using SubscriptionCallback = std::function<void(const ObjectID &)>;
+using SubscriptionFailureCallback = std::function<void(const ObjectID &)>;
+
+/// Interface for the pubsub client.
+class PubsubClientInterface {
+ public:
+  /// Subscribe the object.
+  ///
+  /// \param owner_address Address of the owner to subscribe the object.
+  /// \param object_id The object id to subscribe from the owner.
+  /// \param subscription_callback A callback that is invoked whenever the given object
+  /// information is published.
+  /// \param subscription_failure_callback A callback that is
+  /// invoked whenever the owner is dead (or failed).
+  virtual void SubcribeObject(
+      const rpc::Address &owner_address, const ObjectID &object_id,
+      SubscriptionCallback subscription_callback,
+      SubscriptionFailureCallback subscription_failure_callback) = 0;
+
+  /// Unsubscribe the object.
+  /// NOTE: Calling this method inside subscription_failure_callback is not allowed.
+  /// NOTE: Currently, this method doesn't send a RPC to the pubsub server. It is because
+  /// the client is currently used for WaitForObjectFree, and the coordinator will
+  /// automatically unregister the subscriber after publishing the object. But if we use
+  /// this method for OBOD, we should send an explicit RPC to unregister the subscriber
+  /// from the server.
+  ///
+  /// TODO(sang): Once it starts sending RPCs to unsubscribe, we should start handling
+  /// message ordering. \param owner_address The owner address that it will unsubscribe
+  /// to. \param object_id The object id to unsubscribe.
```

Comment:
```suggestion
  /// message ordering.
  ///
  /// \param owner_address The owner address that it will unsubscribe to.
  /// \param object_id The object id to unsubscribe.
```